subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -365,4 +371,48 @@ protected JavaToolchainQueryService getJavaToolchainQueryService() {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Convention to fall back to the 'destinationDir' output for backwards compatibility with plugins that extend AbstractCompile
+     * and override the deprecated methods.
+     */
+    private static class BackwardCompatibilityOutputDirectoryConvention implements Callable<Directory> {
+        private final AbstractCompile compile;
+        private boolean recursiveCall;
+
+        public BackwardCompatibilityOutputDirectoryConvention(AbstractCompile compile) {
+            this.compile = compile;
+        }
+
+        @Override
+        @Nullable
+        public Directory call() throws Exception {
+            Method getter = GeneratedSubclasses.unpackType(compile).getMethod("getDestinationDir");
+            if (getter.getDeclaringClass() == AbstractCompile.class) {
+                // Subclass has not overridden the getter, so ignore
+                return null;
+            }
+
+            // Subclass has overridden the getter, so call it
+
+            if (recursiveCall) {
+                // Already querying AbstractCompile.getDestinationDirectory()
+                // In that case, this convention should not be used.
+                return null;
+            }
+            recursiveCall = true;
+            File legacyValue;
+            try {
+                // This will call a subclass implementation of getDestinationDir(), which possibly will not call the overridden getter
+                // In the Kotlin plugin, the subclass manages its own field which will be used here.
+                legacyValue = compile.getDestinationDirectory().get().getAsFile();
```

Comment:
This was to support tasks that extended `AbstractCompile` and had their own `getDestinationDir()`.  

We actually need to keep this as `compile.getDestinationDir` to maintain compatibility. 

Can we tell if this is still used by a supported version of the Kotlin plugin? It might be that this workaround isn't necessary anymore.

Reply:
This came up in the prior attempt to remove this property.

It's no longer used in Kotlin > 1.6.21 - [builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_PullRequestFeedback_Trigger/56287468?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_PullRequestFeedback_Trigger/56287468?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true)